### PR TITLE
🌮 MTP-1164: Add checkbox with subsets and search button inside search input field

### PR DESCRIPTION
**Pull request by itemodahl**
MTP-1164

### CODE REVIEW

File: src/scss/components/_list.scss [hunk 1]
```diff
@@ -106,3 +106,25 @@ ol {
     }
   }
 }
+
+.list {
+
+  > li {
+    border-top: 1px solid var(--color-mt-beige-50);
```
Comment (lha047) [inline src/scss/components/_list.scss:113]: Use variable for border width

File: src/scss/components/_list.scss [hunk 2]
```diff
@@ -106,3 +106,25 @@ ol {
...
+    margin-top: 0;
+    padding: 28px;
```
Comment (lha047) [inline src/scss/components/_list.scss:115]: Use spacer size

File: src/scss/components/_list.scss [hunk 3]
```diff
@@ -106,3 +106,25 @@ ol {
...
+    display: flex;
+    justify-content: space-between;
+    align-content: center;
+    &:last-child{
+      border-bottom: 1px solid var(--color-mt-beige-50);
```
Comment (lha047) [inline src/scss/components/_list.scss:120]: Use variabel for border width

File: src/scss/components/_list.scss [hunk 4]
```diff
@@ -106,3 +106,25 @@ ol {
...
+    }
+    &:nth-child(even) {
+      background-color: #fff;
```
Comment (lha047) [inline src/scss/components/_list.scss:123]: Use variable for color

File: src/scss/components/_radio.scss [hunk 5]
```diff
@@ -55,9 +61,32 @@ fieldset {
   legend {
     font-weight: var(--font-weight-bold);
     padding-inline: 0;
+
+    &.border{
+      width: 100%;
+      border-bottom: 1px solid $color-secondary;
```
Comment (lha047) [inline src/scss/components/_radio.scss:67]: Use variable for border width

File: src/scss/components/_radio.scss [hunk 6]
```diff
@@ -55,9 +61,32 @@ fieldset {
...
+    position: relative;
+  }
+
+  .search {
+    background-color: transparent;
+    border: 0 none;
+    cursor: pointer;
+    color: var(--color-mt-blue-medium);
+    position: absolute;
+    right: 10px;
```
Comment (lha047) [inline src/scss/components/_radio.scss:89]: Use variables without px units

File: src/scss/components/_radio.scss [hunk 7]
```diff
@@ -55,9 +61,32 @@ fieldset {
...
+    bottom: 14px;
```
Comment (lha047) [inline src/scss/components/_radio.scss:90]: Use variables without px units

File: src/svelte/components/Tags.svelte [hunk 8]
```diff
@@ -27,3 +28,23 @@
     </li>
   {/each}
 </ul>
+
+<style>
+  .lightblue {
+    background-color: var(--color-mt-rationale-blue);
+    border: 1px solid var(--color-mt-blue-medium);
```
Comment (lha047) [inline src/svelte/components/Tags.svelte:35]: Use variable

File: src/svelte/components/Tags.svelte [hunk 9]
```diff
@@ -27,3 +28,23 @@
     </li>
   {/each}
 </ul>
+
+<style>
```
Comment (lha047) [inline src/svelte/components/Tags.svelte:32]: Move style to scss file

File: src/svelte/components/Tags.svelte [hunk 10]
```diff
@@ -27,3 +28,23 @@
...
+  .lightblue {
+    background-color: var(--color-mt-rationale-blue);
+    border: 1px solid var(--color-mt-blue-medium);
+  }
+
+  .yellowgreen {
+    background-color: var(--color-mt-yellowgreen-lighter);
+    border: 1px solid var(--color-mt-yellowgreen);
```
Comment (lha047) [inline src/svelte/components/Tags.svelte:40]: Use variable

File: src/svelte/components/Tags.svelte [hunk 11]
```diff
@@ -27,3 +28,23 @@
...
+  }
+  .beige {
+    background-color: var(--color-mt-light-beige);
+    border: 1px solid var(--color-mt-beige);
```
Comment (lha047) [inline src/svelte/components/Tags.svelte:44]: Use variable

File: src/svelte/components/Tags.svelte [hunk 12]
```diff
@@ -27,3 +28,23 @@
...
+  }
+
+  .medium {
+    padding: 4px 12px;
```
Comment (lha047) [inline src/svelte/components/Tags.svelte:48]: Use spacer var

File: src/svelte/components/form/CheckboxWithSubSets.svelte [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        } else if (uncheckedChilds.length > 0) {
+          state.element.indeterminate = true
+        } else {
+          state.element.indeterminate = false
+        }
+      }
+    })
+  }
+
+  function mainCategory(mainIndex) {
```
Comment (lha047) [inline src/svelte/components/form/CheckboxWithSubSets.svelte:60]: Param typing. Retur type

File: src/svelte/components/form/CheckboxWithSubSets.svelte [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    } else if (childrenKeys.length > 0) {
+      subSectionValues = subSectionValues.filter(value => !childrenKeys.includes(value))
+      states[mainIndex].children = states[mainIndex].children.map(child => ({
+        ...child,
+        checked: false
+      }))
+    }
+  }
+
+  function subCategory(mainIndex, subCategoryKey) {
```
Comment (lha047) [inline src/svelte/components/form/CheckboxWithSubSets.svelte:77]: Typeing

File: src/svelte/components/form/CheckboxWithSubSets.svelte [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+          </div>
+
+          {#if (states[mainIndex].element.checked || states[mainIndex].element.indeterminate) && listItem.children && listItem.children.length > 0}
+            <ol
+              class="list-unstyled"
+              style="padding-left: 20px"
+              transition:slide={{y: 200, duration: 200}}>
+              {#each listItem.children as subListItem}
+                <li>
+                  <div class="form-control checkbox narrow">
```
Comment (lha047) [inline src/svelte/components/form/CheckboxWithSubSets.svelte:116]: Why do you need the extra div here? Can you use the li-tag

File: src/svelte/components/form/CheckboxWithSubSets.svelte [hunk 16]
```diff
@@ -0,0 +1,138 @@
...
+  }
+</script>
+
+<fieldset>
+  <legend class="border">{title}</legend>
+  {#if states}
+    <ol class="list-unstyled">
+      {#each states as listItem, mainIndex}
+        <li>
+          <div class="form-control checkbox narrow">
```
Comment (lha047) [inline src/svelte/components/form/CheckboxWithSubSets.svelte:97]: Why do you need extra div here? Can you user the li-tag

File: src/svelte/components/form/TextInput.svelte [hunk 17]
```diff
@@ -16,6 +16,7 @@
   export let textOptional: string | undefined
   export let showOptionalText: boolean
   export let hiddenErrorText: string | undefined
+  export let searchButtonText: string | undefined
```
Comment (lha047) [inline src/svelte/components/form/TextInput.svelte:19]: I suggest you create a separate search component. 
This one already have a lot doning on. The search is a separate thing `<input type="search"`

File: stories/components/form/CheckboxWithSubSets.stories.svelte [hunk 18]
```diff
@@ -0,0 +1,92 @@
+<script lang="ts">
```
Comment (lha047) [inline stories/components/form/CheckboxWithSubSets.stories.svelte:1]: I cant see this story

**Review by lha047 — CHANGES_REQUESTED** [review]
Had a quick look :)
Left some comments. 
I think its better to make a separate search component. 
Have a look at using variables.

File: stories/components/form/CheckboxWithSubSets.stories.svelte [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+      {
+        key: 'kosmetikk',
+        displayName: 'Kosmetikk',
+        docCount: 1,
+        children: []
+      }
+    ]
+  }}
+  argTypes={{
+    title: {control: 'text'},
```
Comment (lha047) [inline stories/components/form/CheckboxWithSubSets.stories.svelte:81]: Dont need this if its not shown

File: src/scss/_theme.scss [hunk 20]
```diff
@@ -46,8 +46,10 @@ $icon-size-x-large: 1.875rem; // 30px
 $menu-point-icon-height: 2.125rem; //34px
 $menu-point-icon-width: 2.5rem; // 40px
 
+$spacer-xxxx-small: 0.25rem !default;// 4px
```
Comment (lha047) [inline src/scss/_theme.scss:49]: We should only have the spacers that are defined in Figma.
If you need other spacing start another group of variables

File: src/scss/_theme.scss [hunk 21]
```diff
@@ -46,8 +46,10 @@ $icon-size-x-large: 1.875rem; // 30px
...
 $spacer-xxx-small: 0.375rem !default; // 6px
 $spacer-xx-small: 0.625rem !default; // 10px
+$spacer-x-small: 0.75rem !default; // 12px
```
Comment (lha047) [inline src/scss/_theme.scss:52]: We cant have two variables with the same name

File: src/scss/components/_tags.scss [hunk 22]
```diff
@@ -17,5 +17,23 @@
...
+  .yellowgreen {
+    background-color: var(--color-mt-yellowgreen-lighter);
+    border: var(--border-width-small) solid var(--color-mt-yellowgreen);
+  }
+  .beige {
+    background-color: var(--color-mt-light-beige);
+    border: var(--border-width-small) solid var(--color-mt-beige);
+  }
+
+  .medium {
```
Comment (lha047) [inline src/scss/components/_tags.scss:36]: There is minimal difference between this (medium) and the other kind. 
I think we should select one. Check with design

File: stories/components/List.stories.svelte [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+        type: 'Altinn'
+      },
+      {
+        url: '?content=4',
+        title:
+          'Endre informasjon om Ny matbedrift og/eller import, produksjon og engrossalg av matkontaktmaterialer',
+        type: 'Word'
+      }
+    ],
+    showChapterNumber: true,
```
Comment (lha047) [inline stories/components/List.stories.svelte:33]: Dont need if you dont use

File: stories/components/List.stories.svelte [hunk 24]
```diff
@@ -0,0 +1,48 @@
...
+    disabled: false,
```
Comment (lha047) [inline stories/components/List.stories.svelte:34]: Dont need if you dont use

File: stories/components/List.stories.svelte [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+    disableCss: false
+  }}
+  argTypes={{
+    title: {control: 'text'},
+    content: {control: 'array'},
+    showChapterNumber: {control: 'boolean'},
```
Comment (lha047) [inline stories/components/List.stories.svelte:40]: Dont need if you dont use

File: stories/components/List.stories.svelte [hunk 26]
```diff
@@ -0,0 +1,48 @@
...
+    disabled: {control: 'boolean'},
```
Comment (lha047) [inline stories/components/List.stories.svelte:41]: Dont need if you dont use

File: stories/components/List.stories.svelte [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+    disableCss: {control: 'boolean'}
```
Comment (lha047) [inline stories/components/List.stories.svelte:42]: Dont need if you dont use

File: stories/components/List.stories.svelte [hunk 28]
```diff
@@ -0,0 +1,48 @@
...
+  }} />
+
+<Story name="Normal" let:args let:disableCss let:disabled let:secondary>
+  <h1>Dokument liste</h1>
+  <List listContent={args.content} title={args.title} />
```
Comment (lha047) [inline stories/components/List.stories.svelte:47]: Cant see any list. The `<List>` has a slot not listContent

File: src/scss/components/_list.scss [hunk 29]
```diff
@@ -106,3 +106,25 @@ ol {
     }
   }
 }
+
+.list {
+
+  > li {
+    border-top: 0.0625rem solid var(--color-mt-beige-50);
```
Comment (lha047) [inline src/scss/components/_list.scss:113]: Use variable

File: src/svelte/components/form/CheckboxWithSubSets.test.ts [hunk 30]
```diff
@@ -0,0 +1,79 @@
...
+
+  test('Renders', () => {
+    const {getByText} = render(CheckboxWithSubSets, {
+      options,
+      legend
+    })
+    expect(getByText(legend)).toBeInTheDocument()
+    expect(getByText(`${options[0].displayName} (${options[0].docCount})`)).toBeInTheDocument()
+    expect(getByText(`${options[1].displayName} (${options[1].docCount})`)).toBeInTheDocument()
+  })
```
Comment (lha047) [inline src/svelte/components/form/CheckboxWithSubSets.test.ts:78]: Should include click on the checkbox to display the next level and check that that displays ok

File: src/svelte/components/form/TextInputSearch.svelte [hunk 31]
```diff
@@ -0,0 +1,52 @@
+<script lang="ts">
+  import Label from './Label.svelte'
+  import {createInputAriaDescribedby} from '../../../ts/utils'
+
+  export let value
```
Comment (lha047) [inline src/svelte/components/form/TextInputSearch.svelte:5]:
```suggestion
  export let value: string
```

**Review by lha047 — CHANGES_REQUESTED** [review]
Still some work left here. 
Have a chat with UX about stats styling etc. 
Clean up the stories. The storybook should be useable for others now

File: src/svelte/components/form/Search.svelte [hunk 32]
```diff
@@ -0,0 +1,47 @@
...
+  export let inputClass = ''
+</script>
+
+<fieldset class="input-wrapper">
+  {#if legend}
+    <legend>{legend}</legend>
+  {/if}
+
+  <div class="layout-flex layout-flex-col justify-content-center x" style="--gap: 0">
+    <Label class={labelClass} for={name}>
```
Comment (amundlindstol) [inline src/svelte/components/form/Search.svelte:25]: this is marked as not required in enonic. check for undefined here or require in enonic

File: src/svelte/components/form/CheckboxWithSubSets.svelte [hunk 33]
```diff
@@ -0,0 +1,144 @@
...
+      }
+      return child
+    })
+  }
+</script>
+
+<fieldset>
+  <legend class="border">{legend}</legend>
+  {#if helpText}
+    <p id="">{helpText}</p>
```
Comment (amundlindstol) [inline src/svelte/components/form/CheckboxWithSubSets.svelte:95]: id=""

File: src/svelte/components/form/CheckboxWithSubSets.svelte [hunk 34]
```diff
@@ -0,0 +1,144 @@
...
+              value={listItem.key}
+              aria-describedby={createInputAriaDescribedby(helpText ? name : undefined)}
+              on:change={() => mainCategory(mainIndex)} />
+            <label for={listItem.key}>{`${listItem.displayName} (${listItem.docCount})`}</label>
+          </div>
+
+          {#if (states[mainIndex].element.checked || states[mainIndex].element.indeterminate) && listItem.children && listItem.children.length > 0}
+            <ol
+              class="list-unstyled"
+              style="padding-left: 20px"
```
Comment (amundlindstol) [inline src/svelte/components/form/CheckboxWithSubSets.svelte:117]: change with class="p-l-s" or add p-l-xs in _utils.scss

**Review by amundlindstol — CHANGES_REQUESTED** [review]

File: src/scss/_theme.scss [hunk 35]
```diff
@@ -46,6 +46,7 @@ $icon-size-x-large: 1.875rem; // 30px
 $menu-point-icon-height: 2.125rem; //34px
 $menu-point-icon-width: 2.5rem; // 40px
 
+$spacer-xxxx-small: 0.25rem !default;// 4px
```
Comment (amundlindstol) [inline src/scss/_theme.scss:49]:
```suggestion
```

**Review by amundlindstol — APPROVED** [review]
LGreatTM :D

File: stories/components/form/Search.stories.svelte [hunk 36]
```diff
@@ -0,0 +1,66 @@
...
+  let:helpText>
+  <div use:wrapInShadowDom={disableCss}>
+    <article>
+      <form class="form-layout">
+        <TextInputSearch name="inputfield_1" {label} {searchButtonText} />
+      </form>
+    </article>
+    <article>
+      <form class="form-layout">
+        <TextInputSearch name="inputfield_2" {label} {searchButtonText} labelClass="small" />
```
Comment (lha047) [inline stories/components/form/Search.stories.svelte:37]:
```suggestion
        <TextInputSearch name="inputfield_2" {label} {searchButtonText} />
```

File: stories/components/form/Search.stories.svelte [hunk 37]
```diff
@@ -0,0 +1,66 @@
...
+      </form>
+    </article>
+    <article>
+      <form class="form-layout">
+        <TextInputSearch
+          name="inputfield_3"
+          {label}
+          {helpText}
+          {searchButtonText}
+          labelClass="small" />
```
Comment (lha047) [inline stories/components/form/Search.stories.svelte:57]:
```suggestion
          />
```

**Review by lha047 — CHANGES_REQUESTED** [review]
Good job :D 
Left some more comments :)

**Review by lha047 — CHANGES_REQUESTED** [review]
Nice :D

Tags need to be related to link for screenreaders if used. Remove from story if tags are not to be used

File: stories/components/List.stories.svelte [hunk 38]
```diff
@@ -0,0 +1,52 @@
...
+    ],
+    disableCss: false
+  }}
+  argTypes={{
+    title: {control: 'text'},
+    content: {control: 'array'},
+    disableCss: {control: 'boolean'}
+  }} />
+
+<Story name="Normal" let:args let:disableCss let:disabled let:secondary let:colorScheme>
```
Comment (lha047) [inline stories/components/List.stories.svelte:43]: Remove vars that are not used

File: src/scss/components/_tags.scss [hunk 39]
```diff
@@ -19,17 +19,4 @@
     font-size: $font-size-small;
     height: fit-content;
   }
-  .lightblue {
```
Comment (lha047) [inline src/scss/components/_tags.scss]: The tag colors can stay :)

File: stories/components/Tags.stories.svelte [hunk 40]
```diff
@@ -43,9 +38,6 @@
     <h2>Tags</h2>
     <Tags tags={args.tags} />
     <hr />
-    <h2>Tags med forskjellige farger</h2>
```
Comment (lha047) [inline stories/components/Tags.stories.svelte]: The color tags can stay. We have them in the DS

File: src/scss/components/_tags.scss [hunk 41]
```diff
@@ -17,5 +17,6 @@
     border-radius: $border-radius;
     padding: var(--tag-padding);
     font-size: $font-size-small;
+    height: fit-content;
```
Comment (lha047) [inline src/scss/components/_tags.scss:20]: You should not need this

File: stories/components/Tags.stories.svelte [hunk 42]
```diff
@@ -35,8 +35,10 @@
   <div use:wrapInShadowDom={args.disableCss} />
   <h1>Tags</h1>
   <div class="tags-wrapper">
+    <h2>Tags</h2>
     <Tags tags={args.tags} />
     <hr />
+
```
Comment (lha047) [inline stories/components/Tags.stories.svelte:41]: The color tags can stay. We should have them in the DS

File: src/scss/components/_list.scss [hunk 43]
```diff
@@ -106,3 +106,25 @@ ol {
...
   }
 }
+
+.list {
+
+  > li {
+    border-top: $border-width-small solid var(--color-mt-beige-50);
+    margin-top: 0;
+    padding: var(--spacer-small);
+    display: flex;
```
[the remark this answers is not in the document]
Comment (lha047) [reply src/scss/components/_list.scss:116]: Remove the flex-box styling. We dont need it

**Review by lha047 — CHANGES_REQUESTED** [review]
Nice :D 

Left some comments. The tags story can stay

File: src/scss/components/_list.scss [hunk 44]
```diff
@@ -113,7 +113,6 @@ ol {
     border-top: $border-width-small solid var(--color-mt-beige-50);
     margin-top: 0;
     padding: var(--spacer-small);
-    display: flex;
     justify-content: space-between;
```
Comment (lha047) [inline src/scss/components/_list.scss:116]: Remove `justify-content` and `align-content`

**Review by lha047 — APPROVED** [review]
Nice :D

**Pull request merged**